proposals/jvm-meta-annotations-artifact.md

```diff
@@ -0,0 +1,160 @@
+to all of it's type arguments (recursively).
+
+#### Parameter in @Alias (Option 1)
+The first option is adding `propagateTo: Array<DefaultApplicabilityType>` parameter to `meta.Alias` annotation.
+
+Thus, declaration of `AllParametersAreNullableByDefault` would look like this:
+```kotlin
+@Alias(Nullable::class, propagateTo=[ApplicabilityKind.VALUE_PARAMETER])
+annotation class AllParametersAreNullableByDefault
+```
+
+*Known pros:* 
+- It's nice to have one annotation for all issues.
+- Probably, for someone such form of default qualifiers would look nicer than one suggested by JSR-305.
+
+*Known cons:*
+- It would look a little weird when it's used in cases like the one below:
+```kotlin
+@Alias(Nullable::class)
+@UnderMigration(...)
+annotation class MyNullable
+
+@Alias(MyNullable::class, propagateTo=[ApplicabilityKind.PARAMETER])
+annotation class MyAllParametersAreNullableByDefault
+```
+- Also it might be confusing when one alias (probably with non-empty `propagateTo`)
+references another alias with non-trivial `propagateTo` argument
+
+#### Separate annotation (Option 2)
+Another option would be a separate annotation `meta.ApplyByDefault` with vararg-parameter
+of type `DefaultApplicabilityType`:
+```kotlin
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class ApplyByDefault(val qualifier: KClass<*>, vararg val elements: ApplicabilityKind)
+```
+
+Basically, it should work just the same as `@TypeQualifierDefault`, e.g. it might be used like:
+```kotlin
+@ApplyByDefault(MyNullable::class)
+annotation class MyAllParametersAreNullableByDefault
+```
+
+### Avoidance of @UnderMigration annotation
+After revisiting design for meta-annotations it looks like a natural thing to set up
+a migration status for an annotation with its argument instead of another meta-annotation.
+
+Thus, the idea is to add `migrationStatus` parameter to `meta.Alias` 
+(and to `meta.ApplyByDefault` if we decide to introduce the latter) with default value to be `STRICT`.
+
+This argument should be processed by the compiler in the same way as for [`@UnderMigration`](https://github.com/Kotlin/KEEP/blob/master/proposals/jsr-305-custom-nullability-qualifiers.md#undermigration-annotation) annotation 
+having the same argument value.
+
+### Details on artifact
+
+There is already an artifact called [kotlin-annotations-jvm](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-annotations-jvm)
+that might be the best candidate where the new meta annotations may be placed.
+
+Probably, the annotations that are already there should be moved to a different packages:
+
+- `kotlin.annotations.jvm.MigrationStatus` -> `kotlin.annotations.jvm.meta`
+- `kotlin.annotations.jvm.Mutable`, `kotlin.annotations.jvm.ReadOnly` -> `kotlin.annotations.jvm.collections`
+
+## Remaining questions
+- Should aliasing JSR-305 qualifiers like `javax.annotation.Nonnull` be allowed?
+- What option is the best to choose for supporting [default qualifiers](#default-qualifiers)?
```

Comment:
Option 2 looks cleaner to me, but no strong opinion here. Also something to keep in mind is that JVM seems to generates warning when a library (for example Spring) is using enum annotation attributes that are not in the classpath, which was a use case of JSR 305. Maybe this library will be in the projects classpath since it won't have JSR 305 package issue, but I prefer to mention it in order to be sure this is not an issue for projects that will use this library.

Reply:
What I mean is that the JVM is lenient about meta-annotations not being in the classpath except when these meta-annotations are using enums not in the classpath for their attribute value.

For example if I compile a Spring project without JSR 305 dependency, and if that project is using [`@org.springframework.lang.NonNull`](https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/lang/NonNull.java), no warning when compiling because only `@Nonnull` and `@TypeQualifierNickname` meta-annotations are used.

But if I use [`@org.springframework.lang.Nullable`](https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/lang/Nullable.java), then I get the following warning during compilation: `Warning:java: unknown enum constant javax.annotation.meta.When.MAYBE reason: class file for javax.annotation.meta.When not found` because this annotation is meta-annotated with `@Nonnull(when = When.MAYBE)`.

This issue very painful for such usual use case would not occur if JSR 305 was using `@Nonnull(when = "MAYBE")` or `@Nonnull @Maybe`, so I would be strongly for the new meta-annotation library you will provide avoiding the same mistake as it will allow all libraries to use your meta-annotations without forcing all users of these libraries to have crappy compilation warnings in their logs if they don't have Jetbrains lib in the classpath. IMO this is a key point for wide adoption.

Reply:
NB: AFAIK it only happens when using these annotations in a user's project that depends on Spring project but doesn't have `jsr305.jar` in the classpath. I.e.. when calling a piece of annotated API no warning is emitted.